### PR TITLE
df: Show percentages and (optionally) inode counts

### PR DESCRIPTION
**Pull request by kuzux**

### CODE REVIEW

File: AK/NumberFormat.cpp [hunk 1]
```diff
@@ -25,6 +25,23 @@ DeprecatedString human_readable_size(u64 size)
     return DeprecatedString::formatted("{:.1f} EiB", static_cast<float>(size) / static_cast<float>(EiB));
 }
 
+DeprecatedString human_readable_quantity(u64 quantity)
+{
+    if (quantity < 1 * KiB)
```
Comment (LucasChollet) [inline AK/NumberFormat.cpp:30]: If you want to use power of ten, don't use `KiB` which is in power of two.

File: AK/NumberFormat.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
+/*
```
Comment (LucasChollet) [inline AK/NumberFormat.cpp:1]: This commit isn't atomic. Make another commit to move the definitions to the cpp file.

File: Userland/Utilities/df.cpp [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #include <stdlib.h>
 
 static bool flag_human_readable = false;
```
Comment (LucasChollet) [inline Userland/Utilities/df.cpp:16]: Both variables shouldn't be static

File: AK/NumberFormat.cpp (hunk 1 above)
Comment (kuzux) [reply AK/NumberFormat.cpp:30]: I wanted to use powers of two though, using powers of ten was a option in command line utilities that could be added later

Is it correct to use these constants in this case where the quantity in question is not in bytes but you still want to comparee things with 1024, 1024*1024 etc?

File: AK/NumberFormat.cpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+DeprecatedString human_readable_size(u64 size)
+{
+    static auto suffixes = AK::Array { ""sv, "KiB"sv, "MiB"sv, "GiB"sv, "TiB"sv, "PiB"sv, "EiB"sv };
+    return human_readable_quantity_impl(size, suffixes);
+}
+
+DeprecatedString human_readable_quantity(u64 quantity)
+{
+    static auto suffixes = AK::Array { ""sv, "K"sv, "M"sv, "G"sv, "T"sv, "P"sv, "E"sv };
```
Comment (LucasChollet) [inline AK/NumberFormat.cpp:38]: IMO, we should put a `i` at the end of binary based unit.

File: AK/NumberFormat.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+#include "AK/DeprecatedString.h"
+#include <AK/NumberFormat.h>
+#include <AK/StringView.h>
+
+namespace AK {
+
+// FIXME: Remove this hackery once printf() supports floats.
+static DeprecatedString number_string_with_one_decimal(u64 number, u64 unit, StringView suffix)
+{
+    int decimal = (number % unit) * 10 / unit;
```
Comment (ldm5180) [inline AK/NumberFormat.cpp:17]: I think `int` might not be safe here. `number` and `unit` are both `u64`. Perhaps this should be `u64` or just simply `auto`?

File: AK/NumberFormat.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    return DeprecatedString::formatted("{}.{} {}", number / unit, decimal, suffix);
+}
+
+template<size_t Size>
+static DeprecatedString human_readable_quantity_impl(u64 quantity, const AK::Array<AK::StringView, Size>& suffixes)
```
Comment (ldm5180) [inline AK/NumberFormat.cpp:22]: `1024` seems to be used several times in this function as a magic number. Can this be a named constant?

File: AK/NumberFormat.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+        size_of_unit *= 1024;
+    }
+
+    return number_string_with_one_decimal(quantity, size_of_unit, suffixes.last());
+}
+
+DeprecatedString human_readable_size(u64 size)
+{
+    static auto suffixes = AK::Array { "B"sv, "KiB"sv, "MiB"sv, "GiB"sv, "TiB"sv, "PiB"sv, "EiB"sv };
```
Comment (ldm5180) [inline AK/NumberFormat.cpp:44]: Making this `constexpr` avoids the need to check if it has been initialized each time the function is called.

File: AK/NumberFormat.cpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    return human_readable_quantity_impl(size, suffixes);
+}
+
+DeprecatedString human_readable_quantity(u64 quantity)
+{
+    static auto suffixes = AK::Array { ""sv, "Ki"sv, "Mi"sv, "Gi"sv, "Ti"sv, "Pi"sv, "Ei"sv };
```
Comment (ldm5180) [inline AK/NumberFormat.cpp:50]: Making this `constexpr` avoids the need to check if it has been initialized each time the function is called.

File: AK/NumberFormat.cpp (hunk 5 above)
Comment (kuzux) [reply AK/NumberFormat.cpp:17]: The function previously existed for a while, I just moved it out of line.

For being unsafe: The value is of type `u64` until it is assigned to the variable. At that point it is cast to an `int`. The value being cast to int is always less than 10. Still, just marking it as `auto` would be more consistent with the style. I'll change it to that :)

File: AK/NumberFormat.cpp (hunk 6 above)
Comment (kuzux) [reply AK/NumberFormat.cpp]: It is parameterized now (can be 1024 on 1000)

**Review by LucasChollet — COMMENTED** [review]
If you want to ease the review process, try to minimize each PR. Adding unused options like SI units doesn't help.

File: AK/NumberFormat.cpp [hunk 9]
```diff
@@ -11,7 +11,7 @@
 
 namespace AK {
 
-// FIXME: Remove this hackery once printf() supports floats.
+// NOTE: Using this instead of formatting as a floating point since we do not want the rounding behavior
```
[the remark this answers is not in the document]
Comment (kuzux) [reply AK/NumberFormat.cpp:15]: We had tests that checked for the behavior at the edges (expecting things like 2gb - 1byte rounding to 1.9gb instead of 2.0gb). That behavior seemed reasonable (and somewhat desirable) to me in context of, say, free disk area.

[the remark this answers is not in the document]
Comment (LucasChollet) [reply AK/NumberFormat.cpp:15]: 1. I'm not sure that we want this behavior (IMO it's misleading and just plain false)
2. The formatter isn't consistent enough to do a proper scientific rounding

So we are quite blocked for now. Anyway, considering the behavior we want, we should either:
 - Enhance the formatter, explicitly show that we are rounding down, and using the standard formatter
 - Enhance the formatter, use it and update tests.

But, that's definitely outside the scope of this PR, so please let this FIXME.

**Comment by LucasChollet** [comment]
I won't have the time to review this in the next days.

It just came to my mind: can you change the return type to be a `String` ? Don't mind it if it doesn't work out of the box.

**Comment by kuzux** [comment]
Updating the return type requires a cascading change to `number_string_with_one_decimal`. This involves changing the return type to be `ErrorOr<String>` (since `String::formatted` can fail). While still fairly simple to do, it would be adding way too much work for an already large PR :)

File: Tests/AK/TestNumberFormat.cpp [hunk 10]
```diff
@@ -4,6 +4,7 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include "LibTest/Macros.h"
```
Comment (ADKaster) [inline Tests/AK/TestNumberFormat.cpp:7]: This include isn't required, looks like a clang-tidy addition (if it were required it should be in `<>` not `""`

**Review by ADKaster — CHANGES_REQUESTED** [review]
These changes look good! One final nitpick. 

Also, in the future, make sure that you don't refactor the same file/function multiple times in a patch series. In this case it seems clear that each refactor is adding a unique feature though, so it's fine.
